### PR TITLE
Add aten::nansum and aten::nansum.out

### PR DESCRIPTION
**Pull request by xytintel**
Add aten::nansum and aten::nansum.out.

### CODE REVIEW

**Comment by xytintel** [comment]
image
image

**Comment by xytintel** [comment]
image

File: src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp [hunk 1]
```diff
@@ -36,22 +67,56 @@ struct sum_functor {
...
+};
+
+template <>
+struct NanSumFunctor<c10::complex<at::Half>> {
+  using scalar_t = c10::complex<at::Half>;
+  using acc_t = at::opmath_type<scalar_t>;
+  inline acc_t operator()(acc_t a, acc_t b) const {
+    return a + (at::_isnan(b) ? acc_t{0.} : acc_t{b});
   }
+};
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp:93]: Why we don't do as CUDA does?
image

Comment (fengyuan14) [reply src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp:93]: Why we have specialization for ComplexHalf?

**Comment by fengyuan14** [comment]
Cancel E2E preci since the UT has failed.

File: test/xpu/extended/run_test_with_skip.py [hunk 2]
```diff
@@ -53,6 +53,8 @@
     "test_compare_cpu_div_trunc_rounding_xpu_float16",
     "test_compare_cpu_div_trunc_rounding_xpu_bfloat16",
     "test_compare_cpu_addr_xpu_float16",
+    "test_compare_cpu_nansum_xpu_bfloat16",
+    "test_compare_cpu_nanmean_xpu_bfloat16",
```
Comment (fengyuan14) [inline test/xpu/extended/run_test_with_skip.py:58]: What's the error between CPU ref and XPU impl.

**Comment by xytintel** [comment]
image
image
image
image

File: src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp [hunk 3]
```diff
@@ -36,22 +67,47 @@ struct sum_functor {
...
+  void operator()(TensorIterator& iter) {
+    gpu_reduce_kernel<scalar_t, out_t>(iter, NanSumOps<acc_t, out_t>{});
   }
+};
 
-  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND2(
-      kBool, kComplexHalf, iter.dtype(), "sum_xpu", [&]() {
-        sum_functor<scalar_t>{}(iter);
+template <typename scalar_t>
+struct nansum_functor_complex {
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp:90]:
```suggestion
struct NansumComplexFunctor {
```

File: src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp [hunk 4]
```diff
@@ -36,22 +67,47 @@ struct sum_functor {
...
+    gpu_reduce_kernel<scalar_t, scalar_t>(iter, NanSumOps<acc_t, scalar_t>{});
+  }
+};
+
+void nansum_kernel(TensorIterator& iter) {
+  auto general_dispatcher = [](TensorIterator& iter) {
+    auto dtype = iter.dtype();
+    if (at::isComplexType(dtype)) {
+      AT_DISPATCH_COMPLEX_TYPES_AND(kComplexHalf, dtype, "nansum_xpu", [&]() {
+        nansum_functor_complex<scalar_t>{}(iter);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp:102]:
```suggestion
        NansumComplexFunctor<scalar_t>{}(iter);
```

File: src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp [hunk 5]
```diff
@@ -36,22 +67,47 @@ struct sum_functor {
...
+      });
+    } else {
+      AT_DISPATCH_FLOATING_TYPES(iter.dtype(), "nansum_xpu", [&]() {
+        nansum_functor<scalar_t>{}(iter);
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp:106]:
```suggestion
        NansumFunctor<scalar_t>{}(iter);
```

File: src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp [hunk 6]
```diff
@@ -36,22 +67,47 @@ struct sum_functor {
...
+        });
+  };
+  reduce_dispatch<sum_functor>(iter, general_dispatcher);
+}
+
+template <
+    typename scalar_t,
+    typename acc_t = scalar_t,
+    typename out_t = scalar_t>
+struct nansum_functor {
```
Comment (fengyuan14) [inline src/ATen/native/xpu/sycl/ReduceSumProdKernels.cpp:83]:
```suggestion
struct NansumFunctor {
```

**Review by fengyuan14 — APPROVED** [review]

**Pull request merged**